test/net_http_test.rb

```diff
@@ -0,0 +1,241 @@
+      else
+        yaml_sample_config[sample_env][semian_identifier]
+      end
+    end
+
+    with_semian_options(semian_options_proc) do
+      Net::HTTP.start("localhost", 41_235) do |http|
+        assert_equal yaml_sample_config["development"][http.semian_identifier], http.raw_semian_options
+      end
+      Net::HTTP.start("localhost", 41_234) do |http|
+        assert_equal yaml_sample_config["development"]["http_default"], http.raw_semian_options
+      end
+      assert_equal yaml_sample_config["development"]["http_default"], Semian::NetHTTP.raw_semian_options
+    end
+  end
+
+  def test_adding_custom_errors_work
+    with_custom_errors([::OpenSSL::SSL::SSLError]) do
+      http = Net::HTTP.new("localhost", 41_235)
+      assert http.resource_exceptions.include?(::OpenSSL::SSL::SSLError)
+    end
+  end
+
+  def with_semian_options(options)
+    orig_semian_options = Semian::NetHTTP.raw_semian_options
+    Semian::NetHTTP.raw_semian_options = options
+    yield
+  ensure
+    Semian::NetHTTP.raw_semian_options = orig_semian_options
+  end
+
+  def with_custom_errors(errors)
+    orig_errors = Semian::NetHTTP.custom_exceptions
+    Semian::NetHTTP.custom_exceptions = errors
+    yield
+  ensure
+    Semian::NetHTTP.custom_exceptions = orig_errors
+  end
+
+  def get_subclasses(klass)
+    ObjectSpace.each_object(klass.singleton_class).to_a - [klass]
+  end
+
+  def open_circuit!
+    Net::HTTP.start("localhost", 41_235) do |http|
+      http.read_timeout = 0.1
+      uri = URI('http://localhost:41235/200')
+      http.raw_semian_options[:error_threshold].times do
+        # Cause error error_threshold times so circuit opens
+        @proxy.downstream(:latency, latency: 150).apply do
+          request = Net::HTTP::Get.new(uri)
+          assert_raises Net::ReadTimeout do
+            http.request(request)
+          end
+        end
+      end
+    end
+  end
+
+  def close_circuit!
+    http = Net::HTTP.new("localhost", 41_235)
+    Timecop.travel(http.raw_semian_options[:error_timeout] + 1)
+    # Cause successes success_threshold times so circuit closes
+    http.raw_semian_options[:success_threshold].times do
+      http.get("/200")
+    end
+  end
+
+  class << self
+    attr_accessor :server_thread
+  end
+
+  def start_server_singleton
+    return if self.class.server_thread
+    self.class.server_thread = Thread.new do
+      Thin::Logging.silent = true
+      Thin::Server.start('localhost', 41_234, RackServer)
+    end
+    poll_until_ready(port: 41_234, time_to_wait: 1)
+  end
+
+  def poll_until_ready(port: 41_234, time_to_wait: 1)
+    start_time = Time.now.to_i
+    begin
+      TCPSocket.new('127.0.0.1', port).close
+    rescue Errno::ECONNREFUSED, Errno::ECONNRESET
+      if Time.now.to_i > start_time + time_to_wait
+        raise "Couldn't reach the service on port #{port} after #{time_to_wait}s"
+      else
+        retry
+      end
+    end
+  end
+
+  def after_tests
+    self.class.server_thread.kill
+  end
```

Comment:
I'm not a big fan of this being global for the entire test suite. I'd prefer having a server that's available for the duration of the block, similar to [what we do in `nginx-routing-modules`](https://github.com/Shopify/nginx-routing-modules/blob/master/test/helpers/rack_test_helper.rb).